modules/apps/oauth2-provider/oauth2-provider-service/src/main/java/com/liferay/oauth2/provider/service/impl/OAuth2ScopeGrantLocalServiceImpl.java

```diff
@@ -118,18 +118,22 @@ public Collection<OAuth2ScopeGrant> getOAuth2ScopeGrants(
 		long companyId, String applicationName, String bundleSymbolicName,
 		String accessTokenContent) {
 
-		OAuth2Authorization oAuth2Authorization =
-			_oAuth2AuthorizationLocalService.
-				fetchOAuth2AuthorizationByAccessTokenContent(
-					accessTokenContent);
+		List<OAuth2Authorization> oAuth2Authorizations =
+			oAuth2AuthorizationPersistence.findByC_ATCH(
```

Comment:
This is wrong, you are running db query with hashcode of accessTokenContent, not the accessTokenContent itself.

It must be post verified to confirm the returned entry's accessTokenContent matches the input parameter, due to hash conflict could happen.

Reply:
I am not entirely sure what you mean, as my OAuth2 knowledge is limited.  Are you saying we should perform similar logic as [OAuth2AuthorizationLocalServiceImpl.fetchOAuth2AuthorizationByAccessTokenContent](https://github.com/liferay/liferay-portal/blob/master/modules/apps/oauth2-provider/oauth2-provider-service/src/main/java/com/liferay/oauth2/provider/service/impl/OAuth2AuthorizationLocalServiceImpl.java#L139-L157)?  If so, is it just [the AccessTokenContent check](https://github.com/liferay/liferay-portal/blob/master/modules/apps/oauth2-provider/oauth2-provider-service/src/main/java/com/liferay/oauth2/provider/service/impl/OAuth2AuthorizationLocalServiceImpl.java#L149-L150), or should we also only return the first matching entry?

Or, should we add a new finder to search by CompanyId and AccessTokenContent?  Am I missing your point?

Reply:
You need same check as https://github.com/liferay/liferay-portal/blob/master/modules/apps/oauth2-provider/oauth2-provider-service/src/main/java/com/liferay/oauth2/provider/service/impl/OAuth2AuthorizationLocalServiceImpl.java#L149-L150 for all matching OAuth2Authorization , 2 different accessTokenContent string could have same hashcode, hash conflict is always possible.

Reply:
> 2 different accessTokenContent string could have same hashcode

Makes sense now, thank you!